File: buildSrc/src/main/java/com/doctor/aistring/StringMethodVisitor.java
package com.doctor.aistring;

import org.gradle.api.logging.Logger;
import org.gradle.api.logging.Logging;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.commons.AdviceAdapter;

import java.util.Base64;
import java.util.List;

public class StringMethodVisitor extends AdviceAdapter {
    private final StringEncryptor encryptor;
    private final String key;
    private final String className;
    private final List<String> targetPackages;
    private static final Logger logger = Logging.getLogger(StringMethodVisitor.class);

    public StringMethodVisitor(
            MethodVisitor methodVisitor,
            StringEncryptor encryptor,
            String key,
            List<String> targetPackages,
            int access,
            String name,
            String descriptor,
            String className
    ) {
        super(Opcodes.ASM9, methodVisitor, access, name, descriptor);
        this.encryptor = encryptor;
        this.key = key;
        this.targetPackages = targetPackages;
        this.className = className;
    }

    private boolean shouldEncrypt(String value) {
        if (value == null || value.isEmpty()) {
            return false;
        }
        
        // 只加密指定包名下的代码
        if (this.className != null && targetPackages != null) {
            return targetPackages.stream().anyMatch(pkg -> this.className.contains(pkg));
        }
        
        return false;
    }

    private String encryptString(String value) {
        try {
            byte[] encrypted = encryptor.encrypt(value, key);
            return Base64.getEncoder().encodeToString(encrypted);
        } catch (Exception e) {
            logger.warn("[StringMethodVisitor] 加密失败: {}", e.getMessage());
            return value;
        }
    }

    @Override
    public void visitLdcInsn(Object value) {
        if (value instanceof String && !((String) value).isEmpty() && shouldEncrypt((String) value)) {
            try {
                String processedValue = ((String) value).replace("\n", "\\n");
                String base64Str = encryptString(processedValue);
//                logger.warn("字符串加密: 原始值='{}', 加密后='{}'" , processedValue, base64Str);
                super.visitLdcInsn(base64Str);
                super.visitLdcInsn(key);
                super.visitMethodInsn(
                        Opcodes.INVOKESTATIC,
                        "com/doctor/aistring/AiString",
                        "decrypt",
                        "(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;",
                        false
                );


            } catch (Exception e) {
                logger.warn("[StringMethodVisitor] 处理字符串时发生错误: {}", e.getMessage());
                super.visitLdcInsn(value);
            }
        } else {
            super.visitLdcInsn(value);
        }
    }
}

File: buildSrc/src/main/kotlin/com/doctor/aistring/StringMethodVisitor.kt
package com.doctor.aistring

import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes
import org.objectweb.asm.commons.AdviceAdapter
import java.util.Base64
import org.gradle.api.logging.Logger
import org.gradle.api.logging.Logging

class StringMethodVisitor(
    methodVisitor: MethodVisitor,
    private val encryptor: StringEncryptor,
    private val key: String,
    access: Int,
    name: String,
    descriptor: String
) : AdviceAdapter(Opcodes.ASM9, methodVisitor, access, name, descriptor) {

    private val logger: Logger = Logging.getLogger(StringMethodVisitor::class.java)

    override fun visitLdcInsn(value: Any?) {
        if (value is String && value.isNotEmpty()) {
            try {
                logger.lifecycle("原始字符串: $value")
                logger.lifecycle("使用密钥: $key")
                
                val encrypted = encryptor.encrypt(value, key)
                logger.lifecycle("加密后字节数组: ${encrypted.joinToString(",")}")
                
                val base64Str = Base64.getEncoder().encodeToString(encrypted)
                logger.lifecycle("Base64编码后: $base64Str")
                
                // 加载解密方法所需的参数
                super.visitLdcInsn(base64Str)
                super.visitLdcInsn(key)
                // 调用解密方法
                super.visitMethodInsn(
                    Opcodes.INVOKESTATIC,
                    "com/doctor/aistring/AiString",
                    "decrypt",
                    "(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;",
                    false
                )
                logger.lifecycle("完成字符串处理")
            } catch (e: Exception) {
                logger.error("处理字符串时出错", e)
                super.visitLdcInsn(value)
            }
        } else {
            super.visitLdcInsn(value)
        }
    }
}